src/System.Net.Security/src/System/Net/Security/SNIHelper.cs

```diff
@@ -0,0 +1,251 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Net.Security
+{
+    internal class SNIHelper
+    {
+        public static string GetServerName(byte[] clientHello)
+        {
+            return GetSniFromSslPlainText(clientHello);
+        }
+
+        // https://tools.ietf.org/html/rfc6101#section-5.2.1
+        // SSLPlainText structure:
+        //   - ContentType (1 byte) => 0x16 is handshake
+        //   - ProtocolVersion version (2 bytes)
+        //   - uint16 length
+        //   - opaque fragment[SSLPlaintext.length]
+        private static string GetSniFromSslPlainText(ReadOnlySpan<byte> sslPlainText)
+        {
+            // Is SSL 3 handshake? SSL 2 does not support extensions - skipping as well
```

Comment:
We currently don't "allow" Ssl2 or Ssl3 in .NET Core. It's marked as Obsolete.
https://github.com/dotnet/corefx/blob/master/src/System.Net.Primitives/ref/System.Net.Primitives.cs#L494

But I agree we should probably be correct in parsing it in case it comes from the server.